tests/queries/0_stateless/01653_move_conditions_from_join_on_to_where.sql

```diff
@@ -0,0 +1,38 @@
+DROP TABLE IF EXISTS table1;
+DROP TABLE IF EXISTS table2;
+
+CREATE TABLE table1 (a UInt32, b UInt32) ENGINE = Memory;
+CREATE TABLE table2 (a UInt32, b UInt32) ENGINE = Memory;
+
+INSERT INTO table1 SELECT number, number FROM numbers(10);
+INSERT INTO table2 SELECT number * 2, number * 20 FROM numbers(6);
+
+SELECT '---------Q1----------';
+SELECT * FROM table1 JOIN table2 ON (table1.a = table2.a) AND (table2.b = toUInt32(20));
+EXPLAIN SYNTAX SELECT * FROM table1 JOIN table2 ON (table1.a = table2.a) AND (table2.b = toUInt32(20));
+
+SELECT '---------Q2----------';
+SELECT * FROM table1 JOIN table2 ON (table1.a = table2.a) AND (table2.a < table2.b) AND (table2.b = toUInt32(20));
+EXPLAIN SYNTAX SELECT * FROM table1 JOIN table2 ON (table1.a = table2.a) AND (table2.a < table2.b) AND (table2.b = toUInt32(20));
+
+SELECT '---------Q3----------';
+SELECT * FROM table1 JOIN table2 ON (table1.a = toUInt32(table2.a + 5)) AND (table2.a < table1.b) AND (table2.b > toUInt32(20)); -- { serverError 48 }
+
+SELECT '---------Q4----------';
+SELECT table1.a, table2.b FROM table1 INNER JOIN table2 ON (table1.a = toUInt32(10 - table2.a)) AND (table1.b = 6) AND (table2.b > 20);
+EXPLAIN SYNTAX SELECT table1.a, table2.b FROM table1 INNER JOIN table2 ON (table1.a = toUInt32(10 - table2.a)) AND (table1.b = 6) AND (table2.b > 20);
+
+SELECT '---------Q5----------';
+SELECT table1.a, table2.b FROM table1 JOIN table2 ON (table1.a = table2.a) AND (table1.b = 6) AND (table2.b > 20) AND (10 < 6);
+EXPLAIN SYNTAX SELECT table1.a, table2.b FROM table1 JOIN table2 ON (table1.a = table2.a) AND (table1.b = 6) AND (table2.b > 20) AND (10 < 6);
+
+SELECT '---------Q6----------';
+SELECT table1.a, table2.b FROM table1 JOIN table2 ON (table1.b = 6) AND (table2.b > 20); -- { serverError 403 } 
+
+SELECT '---------Q7----------';
+SELECT * FROM table1 JOIN table2 ON (table1.a = table2.a) AND (table2.b < toUInt32(40)) where table1.b < 1;
+EXPLAIN SYNTAX SELECT * FROM table1 JOIN table2 ON (table1.a = table2.a) AND (table2.b < toUInt32(40)) where table1.b < 1;
+SELECT * FROM table1 JOIN table2 ON (table1.a = table2.a) AND (table2.b < toUInt32(40)) where table1.b > 10;
+
+DROP TABLE table1;
```

Comment:
We need a test to check that this optimization is not applied for outer joins, because `ON` and `WHERE` conditions have different meaning there.

Reply:
Something like this:
```
postgres=# select * from (select 1 x) l full join (select 2 x) r on l.x = r.x and l.x = 1;
 x | x 
---+---
 1 |  
   | 2
(2 rows)

postgres=# select * from (select 1 x) l full join (select 2 x) r on l.x = r.x where l.x = 1;
 x | x 
---+---
 1 |  
(1 row)
```

Reply:
@akuzm
Oh, there is a problem.
In this pr the result of executing :
```

dell123 :) select * from (select 1 x) l full join (select 2 x) r on l.x = r.x and l.x = 1;

┌─x─┬─r.x─┐
│ 1 │   0 │
└───┴─────┘


dell123 :) explain syntax select * from (select 1 x) l full join (select 2 x) r on l.x = r.x and l.x = 1;

┌─explain──────────────┐
│ SELECT               │
│     x,               │
│     r.x              │
│ FROM                 │
│ (                    │
│     SELECT 1 AS x    │
│ ) AS l               │
│ ALL FULL OUTER JOIN  │
│ (                    │
│     SELECT 2 AS x    │
│ ) AS r ON x = r.x    │
│ WHERE x = 1          │
└──────────────────────┘
```

So the state is wrong.  In the previous CH version, it just throw error:
```
dell126 :) select * from (select 1 x) l full join (select 2 x) r on l.x = r.x and l.x = 1;

Received exception from server (version 21.1.1):
Code: 403. DB::Exception: Received from localhost:9002. DB::Exception: Not equi-join ON expression: x = 1. No columns in one of equality side.: While processing x = 1.
```

So I think I should modify the implementation to check whether this is an outer join, if it is an outer join , I will do nothing.  otherwise I will try to optimize it, right?

Reply:
> So I think I should modify the implementation to check whether this is an inner join, if it is not an inner join , I will do nothing. otherwise I will try to optimize it, right?

Yes. Also I suppose we can move conditions for left table in `RIGHT` join and conditions for right table for `LEFT` join. But it's up to you, we can do it later.

Reply:
ok, I will try it.

Reply:
@vdimir it seems that only inner join can move condition in JOIN ON to WHERE.
Because WHERE is executed after join and will remove LEFT/RIGHT join's unmatched result.

Am I right?

Reply:
Yes we don't want to remove unmatched columns. But for `LEFT` join there will not be unmatched rows from right table, so conditions for right table is safe. But it's just an idea, maybe there are better solutions and now we can only do for `INNER`, because it's clear.

Reply:
```
select * from t1 left join t2 on t1.a = t2.a and t2.b = 20;
```
should optimized to 
```
select * from t1 left join (select * form t2 where t2.b = 20) t2 on t1.a = t2.a
```
but not to 
```
select * from t1 left join t2 on t1.a = t2.a where t2.b = 20;
```
Right??? @vdimir

Reply:
Yes, I missed up this when wrote previous comment. Not sure how easy difficult to do it in this way.